clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java

```diff
@@ -2594,27 +2596,20 @@ public void testDropCommitOnBatchExpiry() throws InterruptedException {
         } catch (ExecutionException e) {
             assertTrue(e.getCause() instanceof  TimeoutException);
         }
+
         runUntil(commitResult::isCompleted);  // the commit shouldn't be completed without being sent since the produce request failed.
         assertFalse(commitResult.isSuccessful());  // the commit shouldn't succeed since the produce request failed.
-        assertThrows(TimeoutException.class, commitResult::await);
+        assertThrows(KafkaException.class, commitResult::await);
 
-        assertTrue(transactionManager.hasAbortableError());
-        assertTrue(transactionManager.hasOngoingTransaction());
+        assertTrue(transactionManager.hasFatalBumpableError());
+        assertFalse(transactionManager.hasOngoingTransaction());
         assertFalse(transactionManager.isCompleting());
-        assertTrue(transactionManager.transactionContainsPartition(tp0));
 
-        TransactionalRequestResult abortResult = transactionManager.beginAbort();
-
-        prepareEndTxnResponse(Errors.NONE, TransactionResult.ABORT, producerId, epoch);
-        prepareInitPidResponse(Errors.NONE, false, producerId, (short) (epoch + 1));
-        runUntil(abortResult::isCompleted);
-        assertTrue(abortResult.isSuccessful());
-        assertFalse(transactionManager.hasOngoingTransaction());
-        assertFalse(transactionManager.transactionContainsPartition(tp0));
+        assertThrows(KafkaException.class, () -> transactionManager.beginAbort());
```

Comment:
So, it looks like after this patch, when batch expiration or timeout error, the producer will enter fatal error state after bumping epoch. But before this patch, the we'll abort it and continue the transaction work. Is that right?

Sorry, I didn't realize this situation. This will impact current user behavior, so we need more discussion. I'll ping some experts in this PR, and hope they will help provide comments. 
cc @artemlivshits @ijuma

Reply:
Yes, that is correct. That abort is causing the issue. The producer just assumes that the batches failed, but it is possible that they are still in-flight. When that happens, the abort marker might get processed earlier than the batch. I've seen this in action, and it corrupts the affected partition permanently.

If it is better to keep the producer in a usable state, I can give it a shot. I had one experiment in which I tried keeping the producer usable by increasing the epoch on the client side once. I believe that it is safe to do as the fencing bump will increase the epoch, and the coordinator will never return that to any clients.

Please let me know what you think @ijuma @artemlivshits @showuon

Reply:
I'm wondering, is there a way that we could mitigate this server side? Is it possible to prevent writing the late records after the abort marker? I might be missing something though, so let me know.

Reply:
@jolshan would that mean that each record sent by the producer would have to include the id of a specific transaction (not just the transactional id of the _producer_?)

If the transactional producer sends a transaction id to the coordinator with each record rather than just the producer's id (in which case the coordinator determines whether there is a transaction going on by the order of Start Transaction, Send Record, End Transaction), then this could work. Otherwise, I don't think it's possible to mitigate on the server side.

@showuon I believe that I've seen this bug cause violation of EOS with a transactional producer in the case of a broker failure (90% sure). I'd much rather deal with the producer crashing than deal with incorrect behavior. However, if it's possible to fix this issue without causing a producer crash that would be really nice (:

Reply:
Hmm, entering fatal state doesn't seem desirable. @hachikuji @dajac Do you have thoughts on how to avoid that?

Reply:
@ijuma I think it would be possible to avoid a fatal state, but it would require a client-side epoch bump.
When an IniPid is sent during an ongoing transaction, the coordinator bumps the producer epoch to fence off the current producer. This bumped epoch is never returned to any producers as a valid epoch. This never-exposed epoch could be used by the producer to stay in a usable state.

In short:
epoch=0 -> delivery timeout occurs -> send fencing InitPid -> epoch=1 (on coordinator side) -> increase epoch on client side -> send another InitPid -> safely acquire epoch=2

Since epoch=1 will never be used by another producer, this is a safe operation, and an actual fencing operation (by another producer instance) can be detected.

Reply:
@jolshan not sure about the impact it would have on the overhead of transactions, but having a unique ID per transaction doesn't really seem necessary to me

Reply:
I'm not suggesting a unique transactional ID, but simply bumping the epoch would give us a unique identifier for the transaction in combination with the transaction ID. Again -- this is something I'm considering as a longer term change, and there could be flaws.

Reply:
@jolshan sorry for the confusion, I understood that the uniqueness would be achieved through the epoch bumps - I just don't really see the added value of it.

Reply:
Yeah. No worries. I think what I was thinking of would require a bit more effort -- but the idea is that if the server knew the difference between individual transactions, then it could make better decisions about new writes and markers. (Ie, potentially we could avoid appending the records of an old transaction after a marker for that transaction is appended.) But I also think this idea needs a bit more thought and could require more work than what you are proposing here.